### PR TITLE
Alternator: return full table description on return of DeleteTable

### LINKED ISSUE #11472: Alternator: missing fields in return of DeleteTable
The `DeleteTable` operation in Alternator is supposed to return a `TableDescription` object describing the table which has just been deleted, similar to what `DescribeTable` returns. We should probably just call describe_table() before starting the deletion to build this response.

Our currently implementation only populates the returned `TableDescription` with a few necessary fields - `TableName` and `TableStatus`.

### PR DESCRIPTION
**Pull request by alezzqz**
The DeleteTable operation in Alternator shoudl return a TableDescription object describing the table which has just been deleted, similar to what DescribeTable returns

Fixes #11472

### CODE REVIEW

**Comment by alezzqz** [comment]
Hi @nyh 

>     1. You said "Refs #11472". Doesn't it "Fixes" this issue? If not, we need to say why not - what's still missing (if it's only #5026 that's left then it's fine to say that 11472 is fixed - while leaving 5026 open).
Seems it should be Fixes but not Refs

>     2. If you haven't done so already, please sign the Contributor Agreement in https://www.scylladb.com/open-source-nosql-database/contributor-agreement/. The need for this is described in CONTRIBUTING.md
Already done and got welcome message

>     3. Please write a test for this in test/alternator/test_table.py - that deletes a table and checks that the fields you expect are there. If some are still missing you can leave this test xfailing mentioning #5026, but it's even better to split the test into two - one checking the fields that we already have and the second is about the ones we don't have (and will xfail). Please make sure the new tests pass on DynamoDB (run `test/alternator/run --aws test_table.py::new_test_name`).
Thanks, implemented two tests in test/alternator/test_table.py

File: alternator/executor.cc
Comment (alezzqz) [reply alternator/executor.cc]: > nitpick: since this function always returns a string constant, this can return a std::string_view instead of an sstring.
Done

Comment (alezzqz) [reply alternator/executor.cc]: > Also, if this function is currently used only in executor.cc, I suggest you make it a "static" function (and the table_status class should be defined in executor.cc, not executor.hh).
This function used _proxy member and that's why it wasn't static. I passed _proxy as parameter and made this function static. Done.

File: alternator/executor.cc
Comment (alezzqz) [reply alternator/executor.cc]: This function used _proxy class member that's why it wasn't static. But I've used it as parameter and made it static. Done.

File: alternator/executor.hh
Comment (alezzqz) [reply alternator/executor.hh:62]: Hm. Since we use this enum in **fill_table_description** parameter we should keep this declaration in executor.hh, right?

File: alternator/executor.cc
Comment (alezzqz) [reply alternator/executor.cc]: fixed this comment

File: alternator/executor.cc
Comment (alezzqz) [reply alternator/executor.cc]: Researched this. True. Moved this comment to the end of **fill_table_description**

**Comment by alezzqz** [comment]
>Please make sure the new tests pass on DynamoDB (run `test/alternator/run --aws test_table.py::new_test_name`).
I checked all these new tests but unfortunately couldn't run it using --aws option. There are boto3 runtime errors occur like:
`botocore.exceptions.NoRegionError: You must specify a region`
and
`unable to locate credentials`
so it's impossible to run any test in this mode. Maybe I have to configure something, don't know

**Comment by nyh** [comment]
> > Please make sure the new tests pass on DynamoDB (run `test/alternator/run --aws test_table.py::new_test_name`).
> 
> I checked all these new tests but unfortunately couldn't run it using --aws option. There are boto3 runtime errors occur like: `botocore.exceptions.NoRegionError: You must specify a region` and `unable to locate credentials` so it's impossible to run any test in this mode. Maybe I have to configure something, don't know

Yes, when you use the "--aws" option it uses the "real" DynamoDB on AWS, and you need to configure your installation on which region to contact and what keys to use so AWS will charge you. I.e., create in `$HOME/.aws` the file `config`:
```
[default]
region = us-east-1
```
and `credentials`:
```
[default]
aws_access_key_id = *******
aws_secret_access_key = *************
```

This is explained in `test/alternator/README.md`.

**Comment by nyh** [comment]
> Hi @nyh
> 
> > ```
> > 1. You said "Refs #11472". Doesn't it "Fixes" this issue? If not, we need to say why not - what's still missing (if it's only #5026 that's left then it's fine to say that 11472 is fixed - while leaving 5026 open).
> > ```
> 
> Seems it should be Fixes but not Refs

Then please change both the commit message of the individual commit, and the top comment in this pull request which you should edit because it will become the cover letter of the PR (i.e., a commit message of the merge).

File: alternator/executor.cc
Comment (nyh) [reply alternator/executor.cc]: Please combine the two patches into one so I can review the final result, instead of seeing something wrong in the first patch and then a fix in a second patch.

File: alternator/executor.cc
Comment (nyh) [inline alternator/executor.cc:426]: This can now be a static (file-local) function and doesn't need to be a static (class-local) "executor" method.

**Comment by alezzqz** [comment]
> Then please change both the commit message of the individual commit, and the top comment in this pull request which you should edit because it will become the cover letter of the PR (i.e., a commit message of the merge).

Done. Changed both individual commit and pull request message.

File: test/alternator/test_table.py
Comment (nyh) [inline test/alternator/test_table.py:562]: After you call create_table() and before you can delete it, you need to wait for the table to have been created (this is not important on Alternator, but it's important on DynamoDB - where this CREATING phase can last many seconds). See an example in the create_and_delete_table() function.

You can use the create_test_table() utility function, which picks a unique table name (using unique_table_name() instead of 'alternator_test') and waits for the creation to complete.

**Comment by alezzqz** [comment]
> @alezzqz by the way, if it's too difficult for you to run the tests on AWS DynamoDB (which requires a real AWS account), I can run them for you, no problem.

I'd appreciate it so much. As far as I understand it seems I should have AWS credentials and it can take a while to get them...

File: test/alternator/test_table.py
Comment (nyh) [inline test/alternator/test_table.py:560]: Please use a unique table name (e.g., "unique_table_name()") instead of a fixed string. These fixed strings can become a problem if you want to repeat the same test multiple times, for example, and didn't wait for the table to be deleted completely before starting the next iteration.

File: test/alternator/test_table.py
Comment (nyh) [inline test/alternator/test_table.py:600]: Please mention "#5026" in the reason (the "reason" means why we expect the test to fail today). The phrase "shouldn't contain these fields" is confusing because actually it should contain these fields...

File: test/alternator/test_table.py
Comment (nyh) [inline test/alternator/test_table.py:605]: Again, if you run this on AWS you'll see you can't call delete immediately after create_table. Also here you can use the create_test_table() utility function which waits for the creation to complete.

**Review by nyh — CHANGES_REQUESTED** [review]
Looks good. I left some relatively-minor comments and requests.

**Comment by nyh** [comment]
> > @alezzqz by the way, if it's too difficult for you to run the tests on AWS DynamoDB (which requires a real AWS account), I can run them for you, no problem.
> 
> I'd appreciate it so much. As far as I understand it seems I should have AWS credentials and it can take a while to get them...

No problem. This feature is very simple so I can also "imagine" how AWS DynamoDB will behave (e.g., as I commented in my review, I know it will have a problem with an attempt to delete a table without waiting for its creation first), but in general, and in the long run, the best way to develop in Alternator is to find a way to run your tests on the "real" DynamoDB and see how it really behaves - instead of how you think it should be have given the documentation. DynamoDB's documentation is usually quite accurate - but not always and some things aren't obvious to understand in the documentation without trying them out.

In any case, when you have a version that I'm happy with, I'll also run the tests myself (against Alternator *and* AWS) before committing the patch, so nothing to worry about.

File: alternator/executor.cc
Comment (alezzqz) [reply alternator/executor.cc]: Done. Combined commits to the single patch

File: alternator/executor.cc
Comment (alezzqz) [reply alternator/executor.cc]: Done. Also moved table_status enum to executor.cc

File: test/alternator/test_table.py
Comment (alezzqz) [reply test/alternator/test_table.py:562]: Thanks, I used test_table.py:create_and_delete_table function as example. Done.

File: test/alternator/test_table.py
Comment (alezzqz) [reply test/alternator/test_table.py]: Done

File: test/alternator/test_table.py
Comment (alezzqz) [reply test/alternator/test_table.py]: Done

File: test/alternator/test_table.py
Comment (alezzqz) [reply test/alternator/test_table.py:608]: the same like for previous test: used **create_and_delete**_table function as example. Done

File: alternator/executor.cc
Comment (alezzqz) [reply alternator/executor.cc]: btw, is it better and more convenient for you to combine commits after every change in pull request?

Comment (nyh) [reply alternator/executor.cc]: You don't have to combine all the commits into one, but yes, I prefer not to see commit 2 fixing commit 1 - it makes it harder to review (I don't remember what I reviewed last time, and my comment is still on commit 1), and makes the "git log" later ugly.
In a larger change that can be split into multiple commits on a logical basis, that's even better, but this wasn't the case here.

Comment (alezzqz) [reply alternator/executor.cc]: Okay, I understand. Will try to do more convenient for you to review. Thanks.

File: test/alternator/test_table.py
Comment (nyh) [inline test/alternator/test_table.py:565]: nitpick: you did not have to wait for the tale_not_exists event. You care about the response from table.delete() (which you saved in "got") - you don't care when the actual deletion finishes.

**Review by nyh — CHANGES_REQUESTED** [review]
Thanks. Looks mostly good, except that test_delete_table_description doesn't pass on DynamoDB: it fails the "assert 'CreationDateTime' in got" test. This is surprising because https://docs.aws.amazon.com/amazondynamodb/latest/APIReference/API_DeleteTable.html suggests it should exist in the response of DeleteTable... But it turns out that it doesn't! 
Here is what I get in "got" in that test on AWS DynamoDB:

> {'TableName': 'alternator_Test_1664802526950', 'TableStatus': 'DELETING', 'ProvisionedThroughput': {'NumberOfDecreasesToday': 0, 'ReadCapacityUnits': 0, 'WriteCapacityUnits': 0}, 'TableSizeBytes': 0, 'ItemCount': 0, 'TableArn': 'arn:aws:dynamodb:us-east-1:797456418907:table/alternator_Test_1664802526950', 'TableId': '530cc625-8448-4eab-80b6-294daf9a7768', 'BillingModeSummary': {'BillingMode': 'PAY_PER_REQUEST', 'LastUpdateToPayPerRequestDateTime': datetime.datetime(2022, 10, 3, 16, 8, 47, 612000, tzinfo=tzlocal())}}

Maybe you should just drop the assert from the test and leave a comment that although the documentation suggest it should exist, it isn't necessary and DynamoDB doesn't return it?

File: test/alternator/test_table.py
Comment (nyh) [inline test/alternator/test_table.py:602]: Please remove this "strict=True". I want to be able to run this test on AWS DynamoDB, and see it XPASS, not FAIL. 
Also, If one day we fix this bug in Alternator and forget to remove this xfail, we will later notice an XPASS and can remove the xfail - it's not a good reason (I think) to fail the test.

**Review by nyh — CHANGES_REQUESTED** [review]

**Comment by alezzqz** [comment]
> Maybe you should just drop the assert from the test and leave a comment that although the documentation suggest it should exist, it isn't necessary and DynamoDB doesn't return it?

Agree. That's a best solution in the current situations. Just commented that assertion and wrote a comment.

File: test/alternator/test_table.py
Comment (alezzqz) [reply test/alternator/test_table.py]: Done

File: test/alternator/test_table.py
Comment (alezzqz) [reply test/alternator/test_table.py]: Done

**Comment by nyh** [comment]
> CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/1441/

CI failed, I couldn't figure out why :-( I'll restart.

**Review by nyh — APPROVED** [review]
Looks good. I just have a minor comment, and want to see CI passing on this version.

**Comment by nyh** [comment]
> Looks good. I just have a minor comment, and want to see CI passing on this version.

If you haven't rebased on a recent "master" recently, maybe you should do it now - maybe that's the problem with the CI? Anyway, I also restarted the CI, so we'll see.

File: alternator/executor.cc
Comment (alezzqz) [reply alternator/executor.cc]: Done

**Comment by alezzqz** [comment]
> > Looks good. I just have a minor comment, and want to see CI passing on this version.
> 
> If you haven't rebased on a recent "master" recently, maybe you should do it now - maybe that's the problem with the CI? Anyway, I also restarted the CI, so we'll see.

I've rebased on latest master yesterday before updating the commit. Checked build in release/dev/debug, all units test passed locally. Unfortunately I have no access to Jenkins so couldn't see the problem (

**Comment by nyh** [comment]
> > > Looks good. I just have a minor comment, and want to see CI passing on this version.
> > 
> > 
> > If you haven't rebased on a recent "master" recently, maybe you should do it now - maybe that's the problem with the CI? Anyway, I also restarted the CI, so we'll see.
> 
> I've rebased on latest master yesterday before updating the commit. Checked build in release/dev/debug, all unit tests passed locally. Unfortunately I have no access to Jenkins so couldn't see the problem (

Thanks. I looked at the CI (@mykaul @avikivity isn't it visible outside ScyllaDB the company?) and it turns out that we hit two known flaky tests, unrelated to your patch :-( I restarted the run yet again.

**Comment by alezzqz** [comment]
> > > > Looks good. I just have a minor comment, and want to see CI passing on this version.
> > > 
> > > 
> > > If you haven't rebased on a recent "master" recently, maybe you should do it now - maybe that's the problem with the CI? Anyway, I also restarted the CI, so we'll see.
> > 
> > 
> > I've rebased on latest master yesterday before updating the commit. Checked build in release/dev/debug, all unit tests passed locally. Unfortunately I have no access to Jenkins so couldn't see the problem (
> 
> Thanks. I looked at the CI (@mykaul @avikivity isn't it visible outside ScyllaDB the company?) and it turns out that we hit two known flaky tests, unrelated to your patch :-( I restarted the run yet again.

I'd be happy if I had just read only access to CI to see what's happening there )

**Comment by alezzqz** [comment]
seems everything is fine now?

**Comment by alezzqz** [comment]
I still can't merge this PR even after approvement: not authorized to do that. Maybe I should have some additional permissions?

**Comment by nyh** [comment]
Sorry, I'll merge it now.
Only designated "maintainers" of the project can actually merge things - approvals and CI are important but do not allow other people to merge.

**Review by nyh — APPROVED** [review]

**Comment by nyh** [comment]
I was about to merge this, but then I noticed that the test `test_delete_table_description` fails when run on AWS DynamoDB:

```
test/alternator/run --aws test_table.py::test_delete_table_description
```

It seems that contrary to what the test assumes and the documentation https://docs.aws.amazon.com/amazondynamodb/latest/APIReference/API_DeleteTable.html says, DynamoDB does NOT return several of the fields in the DeleteTable response: It seems it does not return `KeySchema`, `AttributeDefinitions`, and `CreationDateTime`.

I don't know why DynamoDB doesn't have these parts of the response, but maybe we shouldn't have them either in this case, even if it's easy for us to return?
Alternatively if we choose to keep them (but... why?), the test should be modified to check less, and pass on both DynamoDB and Alternator.

**Comment by nyh** [comment]
It looks like one of my hands typed "I was about to merge it", while the other hand merged it ;-)
I'll open a new issue about what's wrong now.

**Comment by nyh** [comment]
Created https://github.com/scylladb/scylladb/issues/14132

**Pull request closed without merging**
